packages/ptokens-assets-evm/src/ptokens-evm-provider.ts

```diff
@@ -205,49 +232,91 @@ export class pTokensEvmProvider implements pTokensAssetProvider {
         return false
       }
     }, _pollingTime)
-    return receipt.transactionHash.toString()
+    if (!receipt) throw new Error('Polling function stopped unexpectedly. No receipt found')
+    return receipt.transactionHash.toString() // not clear why TransactionReceipt
   }
 
   protected async _pollForHubOperation(
     _hubAddress: string,
     _eventName: EVENT_NAMES,
     _fromBlock: bigint,
     _operationId: string,
+    _onInterim: boolean,
     _pollingTime = 1000,
-  ) {
-    let log
+  ): Promise<Log> {
+    let log: Log | undefined
     // eslint-disable-next-line @typescript-eslint/no-unsafe-call
     await polling(async () => {
       try {
-        const logs = await this._publicClient.getLogs({
-          fromBlock: _fromBlock,
-          address: stringUtils.addHexPrefix(_hubAddress),
-          events: events,
-        })
+        log = undefined
+        const logs = _onInterim ?
+          await this._interimClient.getLogs({
+            fromBlock: _fromBlock,
+            address: stringUtils.addHexPrefix(_hubAddress),
+            events: pNetworkHubAbi.filter((_fun) => _fun.type === 'event' && _fun.name === _eventName.toString()),
+          }) :
+          await this._publicClient.getLogs({
+            fromBlock: _fromBlock,
+            address: stringUtils.addHexPrefix(_hubAddress),
+            events: pNetworkHubAbi.filter((_fun) => _fun.type === 'event' && _fun.name === _eventName.toString()),
+          })
         log = logs.find((_log) => getOperationIdFromLog(_log) === _operationId)
         if (log) return true
         return false
-      } catch (_err) {
+      } catch (_err) {  
         return false
       }
     }, _pollingTime)
+    if (!log) throw new Error('Polling function stopped unexpectedly. No log found')
     return log
   }
 
-  monitorCrossChainOperations(_hubAddress: string, _operationId: string) {
+  monitorCrossChainOperations(_hubAddress: string, _interimHubAddress: string, _operationId: string) {
```

Comment:
I would move this function to the evm provider.
Also, I would remove _interimHubAddress.
You call monitorCrossChainOperation(interimHub),
return something
then call monitorCrossChainOperation(destinationHub).
monitorCrossChainOperation(interimHub) could be easily called by pTokensSwap after makeContractSend, using the evmProvider (on polygon) supplied to the pTokensSwap constructor.